pkg/lib/filesystem/ignore.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024 The KitOps Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package filesystem
+
+import (
+	"fmt"
+	"kitops/pkg/artifact"
+	"kitops/pkg/lib/constants"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/moby/patternmatcher"
+	"github.com/moby/patternmatcher/ignorefile"
+)
+
+type IgnorePaths interface {
+	Matches(path, layerPath string) (bool, error)
+	HasExclusions() bool
+}
+
+func NewIgnore(contextDir string, kitfile *artifact.KitFile) (IgnorePaths, error) {
+	kitIgnorePaths, err := readIgnoreFile(contextDir)
+	if err != nil {
+		return nil, err
+	}
+	kitIgnorePaths = append(kitIgnorePaths, constants.DefaultKitfileNames()...)
+	kitIgnorePaths = append(kitIgnorePaths, constants.IgnoreFileName)
+	kitIgnorePM, err := patternmatcher.New(kitIgnorePaths)
+	if err != nil {
+		return nil, fmt.Errorf("invalid %s file: %w", constants.IgnoreFileName, err)
+	}
+
+	layerPaths := layerPathsFromKitfile(kitfile)
+	return &ignorePaths{
+		ignoreFileMatcher: kitIgnorePM,
+		layers:            layerPaths,
+	}, nil
+}
+
+type ignorePaths struct {
+	ignoreFileMatcher *patternmatcher.PatternMatcher
+	layers            []string
+}
+
+func (pm *ignorePaths) Matches(path, layerPath string) (bool, error) {
+	ignoreFileMatches, err := pm.ignoreFileMatcher.MatchesOrParentMatches(path)
+	if err != nil {
+		return false, err
+	}
+	if ignoreFileMatches {
+		return true, nil
+	}
+	// ignore file doesn't exclude the current path, check if it should be excluded
+	// since it's included in another layer
+	for _, layer := range pm.layers {
```

Comment:
Feels like there should be a guard somewhere here to skip the currently packaged layer.

Reply:
How do you mean? This check is a little weird but I think it hits the main issue -- given a Kitfile
```yaml
manifestVersion: 1.0.0
package:
  name: test-intersecting
model:
  name: test-main
  path: main-dir
datasets:
  - name: test-intersect
    path: main-dir/intersect-dir
```
there are two cases:

* while processing the model (`main-dir` path), we want to ignore all paths prefixed with `main-dir/intersect-dir`
* while processing the dataset (`main-dir/intersect-dir` path), we _don't_ want to ignore all paths prefixed `main-dir` since that would ignore the whole layer

Worth noting that this was broken until the latest commit if you used `./main-dir` in the Kitfile, since the clean version of that path is `main-dir`, which breaks the prefix check.

Reply:
yes. I was using `./main-dir`